Add Activity entity in Subgraph

george-maroun opened the pull request:
- Create an Activity entity
- Add code in governance handler to create an save a new activity
- Add stake and unstake handlers in glow.ts
- Create tests to ensure activities are created when user performs actions ie vote, stake/unstake, veto, create proposal, reject/ratify a proposal

UPDATE: also add totalStakedGlow to User and properly handle stake and unstake events + add tests in glow.test.ts

george-maroun committed:
Add Activity entity

george-maroun commented on schema.graphql:
```diff
@@ -176,6 +176,23 @@ type NominationSpend @entity(immutable: true) {
+  proposal: Proposal
+}
+
 type User @entity {
```
Should we add a property for Activities in the User entity?

0xSimbo replied:
Yes, you want to attach it as a relationship and then in activities reference the User.

george-maroun commented on schema.graphql:
```diff
@@ -176,6 +176,23 @@ type NominationSpend @entity(immutable: true) {
+  activityType: ActivityType!
+  timestamp: BigInt!
+  transactionHash: Bytes!
+  proposal: Proposal
```
For ratification and rejection vote breakdown, do you think i should add an optional property of type MostPopularProposalVoteBreakdown
it can be assigned in ratification and rejection activities

george-maroun committed:
WIP testing stake handler

george-maroun committed:
WIP testing

george-maroun committed:
Tests pass

george-maroun commented on src/glow.ts:
```diff
@@ -23,3 +24,52 @@ function _handleTransferToCarbonCreditAuction(event: GlowTransferEvent): void {
+  let userAddress = event.params.user.toHexString();
+  let stakeAmount = event.params.amount;
+
+  // TODO: Maybe keep track of the stake amount by user
```
should we keep track of the stake amount by user?

0xSimbo replied:
Are you referencing the total amount staked by a user? We could have this in the subgraph.

george-maroun commented on src/shared/createActivity.ts:
```diff
@@ -0,0 +1,62 @@
+  transactionHash: string,
+  logIndex: string
+): string {
+  return "activity-" + activityType + "-" + userAddress + "-" + transactionHash + "-" + logIndex;
```
is this too long ?

george-maroun replied:
I’m not sure i need the logIndex. can a user perform the same activity on the same transaction multiple times? 
Also is logIndex appropriate or should i use nounce?

0xSimbo replied:
The user can technically perform the same activity on the same transaction multiple times. What you can do is hash all of the fields together to make an identifier so it's more compact. LogIndex is necessary.

george-maroun replied:
Okay I kept "activity" and activityType as a prefix and hashed the rest. I am actually using a substring of the hash that is 24 chars long. The p(collision) is negligible

george-maroun committed:
Cleanup

george-maroun commented on src/governance.ts:
```diff
@@ -72,6 +75,15 @@ export function vetoCouncilElectionOrSlashHandler(
+    event.params.proposalId.toString(),  // proposalId
+    null, // votes
+    null, // glowAmount
+  );
```
One thing I’m not 100 % sure of is if I’m creating activities with the correct type in this file @0xSimbo

0xSimbo replied:
I think that the wording on the activity is a bit confusing. You might want to change the name of the activity to VetoProposalCreation because VetoCouncil members can veto proposals which is a completely different action. You can also get the number of nominations used on the proposal from the event I believe.

```/**
     * @notice Emitted when a Veto Council Election or Slash proposal is created
     * @param proposalId the id of the proposal
     * @param proposer the address of the proposer
     * @param oldAgent the address of the old agent
     * @param newAgent the address of the new agent
     * @param slashOldAgent whether or not to slash the old agent
     * @param nominationsUsed the amount of nominations used
     */
    event VetoCouncilElectionOrSlash(
        uint256 indexed proposalId,
        address indexed proposer,
        address oldAgent,
        address newAgent,
        bool slashOldAgent,
        uint256 nominationsUsed
    );```
    
    Votes are different than nominations used on propsals. Votes come from ratify or reject activities while nomination spent comes from creating proposals and using nominations on proposals.

0xSimbo replied:
You can also find the rest of the events here:
https://github.com/glowlabs-org/glow-contracts/blob/dev/src/interfaces/IGovernance.sol

george-maroun replied:
Sounds good! do you think the type of this action could just be "Create" since the user is creating a proposal? (to be consistent with the other action types)

And on the frontend, we can get the proposal type from proposal id:
 proposalType: 'VetoCouncilElectionOrSlash',
and in the UI say something like "0x123abc created a Veto Council Election or Slash proposal using x nominations"

otherwise i can make all the actions a bit more descriptive. e.g.: instead of "Create" use "GrantProposalCreation" etc.

george-maroun commented:
PTAL @0xSimbo

george-maroun committed:
Revert: remove nominations used from activity schema

0xSimbo commented on build/subgraph.yaml:
```diff
@@ -18,7 +18,7 @@ dataSources:
       abis:
         - name: GCC
-          file: GCC/abis/GCC.json
+          file: GCC/GCC.json
```
Did the abi folder change?

0xSimbo replied:
Didnt realize i was looking at the build folder.

0xSimbo commented on src/governance.ts:
```diff
@@ -390,6 +458,15 @@ export function proposalVetoedHandler(event: ProposalVetoedEvent): void {
   mostPopularProposal.vetoer = event.params.vetoer.toHexString();
   mostPopularProposal.isVetoed = true;
   mostPopularProposal.save();
+
+  createActivity(
+    event,
+    "Veto",
+    event.params.vetoer.toHexString(),
+    event.params.proposalId.toString(),  // proposalId
+    null, // votes
+    null, // glowAmount
+  );
```
"Veto" has the same word as the vetoCouncilElection handler which is confusing.

george-maroun committed:
Address comments

george-maroun committed:
add totalStakedGlow to User and properly handle stake and unstake events

george-maroun committed:
Cleanup

george-maroun commented:
Addressed your comments and added a property for totalStakedGlow on User + handle stake and unstake events appropriately.

Here's what changed since your review: 
- glow.ts: update totalStakedAmount when user stakes or unstakes
- glow-test.ts: add tests for stake update handlers
- governance.ts: change "Veto" to "Create" in vetoCouncilElectionOrSlashHandler + Add nominationsUsed in "Create" and "Nominate" type activities.
- getOrCreateUser.ts: add a line to initialize totalStakedAmount: BigInt
- createActivity.ts: hash the activity key to make it shorter
- schema.graphql: Add totalStakedAmount: BigInt!  property to User + Add nominationsUsed property to Activity and "Nominate" in ActivityType enum

PTAL @0xSimbo

george-maroun committed:
Add activity in nominationsUsedOnProposalHandler, + track nominationsUsed

george-maroun committed:
Cleanup

george-maroun commented on schema.graphql:
```diff
@@ -186,9 +209,11 @@ type User @entity {
   totalUSDCEffect: BigInt!
   nonceSeperator: BigInt!
   totalImpactPoints: BigInt!
+  totalStakedGlow: BigInt!
```
The totalStakedGlow property changes when the stake or unstaked event handlers are invoked. but if the user already has staked glow then the original amount will be 0 in the User table @0xSimbo
